documentation/api/query/v4/upgrading-from-v3.markdown

```diff
@@ -0,0 +1,125 @@
+---
+title: "PuppetDB 3.0 » API » v4 » Upgrading from v3"
+layout: default
+canonical: "/puppetdb/latest/api/query/v4/upgrading-from-v3.html"
+---
+
+This document describes new features of the PuppetDB v4 API relative to v3,
+and the actions developers will need to take to make their code compliant. This
+is not a changelog for PuppetDB 3.0, since some of the changes described are
+present in 2.2.x.
+
+The major changes since v3 (encompassing some that are already available on v4
+in 2.2.2 and some that are not yet released) break down as follow:
+
+### CHANGES THAT MAY BREAK YOUR CODE
+
+* All previously dash-separated field names (e.g receive-time), subquery
+  operators (e.g select-facts), and query parameters (e.g order-by) are now
+  underscore-separated. This change does not apply to dash-separated endpoint
+  names such as `aggregate-event-counts`.
+
+* We have renamed the "name" key of the catalogs endpoint to "certname", for
+  consistency with other endpoints.
+
+* The metrics endpoint has been split off into a separate service, and
+  reversioned at v1. This means if you are currently accessing metrics at
+  http://localhost:8080/v3/metrics/mbeans, you will now access them at
+  http://localhost:8080/metrics/v1/mbeans and so on according to the documentation
+  here: https://docs.puppetlabs.com/puppetdb/master/api/metrics/v1/index.html
+
+* The v4 facts endpoint will return proper JSON rather than stringified JSON
+  under the `value` field in the case of a structured fact. This is observable
+  in current stable (PDB 2.2.x or PE 3.7) with the v4 API.
+
+* The v4 catalogs endpoint no longer produces a `metadata` field or the
```

Comment:
yeah, this doc is highlighting differences between v3 and v4. Some released changes are included.